### PR TITLE
Add tests for redirect middleware & refactoring

### PR DESCRIPTION
**Pull request by JasonChong96**
## Problem

As described in #70, there is a lack of automated tests for this repository. Much of the codebase remains highly coupled.

Closes [#70]

## Solution

- Add unit test cases for the redirect middleware and its' interface with the data layer.
- Refactor `login.ts` to use dependency injection

**Improvements**:

- There are now some automated test cases for the redirect endpoint
- Testing libraries and frameworks such as jest have been integrated 
- `login.ts` now uses dependency injection. Functions are also separated from routing logic.

## TODOs

- [x] Add unit tests for `redirect.ts`
- [x] Add unit tests for `repositories/url.ts` and `cache/url.ts`
- [x] Integrate unit tests with TravisCI
- [x] Refactor `login.ts` to use dependency injection
- [x] Rebase with latest release

## Tests

- Run `npm test` in the repository's root folder. Should pass with zero exit code.
- Change the code coverage threshold in `jest.config.js` to 100 and run `npm test` again. The test should fail with a non-zero exit code.

**New dev dependencies**:

- `node-mocks-http`: Provides mocks for express request and response objects
- `jest`: A testing framework for node.js
- `ts-jest`: Provides typescript compatibility for `jest`
- `@types/jest`: Typescript types for `jest`
- `redis-mock`: Mock `redis` implementation for unit testing
- `@tyles/redis-mock`: Typescript types for `redis-mock`
- `sequelize-mock`: Mock `sequelize` implementation for unit testing
- `eslint-plugin-jest`: Adds `jest` support for `eslint` and ensures tests are of good quality
- `sinon`: Standalone test spies, stubs and mocks for JavaScript
- `@types/sinon`: Typescript types for `sinon`

**Additional Notes:**

Minimum code coverage for this repository is now set to 10%.

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -28,42 +28,39 @@
...
-    "noImplicitThis": true,                /* Raise error on 'this' expressions with an implied 'any' type. */
-    "alwaysStrict": true,                  /* Parse in strict mode and emit "use strict" for each source file. */
-
+    "noImplicitAny": true,                    /* Raise error on expressions and declarations with an implied 'any' type. */
+    "strictNullChecks": true,                 /* Enable strict null checks. */
+    "strictFunctionTypes": true,              /* Enable strict checking of function types. */
+    "strictBindCallApply": true,              /* Enable strict 'bind', 'call', and 'apply' methods on functions. */
+    "strictPropertyInitialization": true,     /* Enable strict checking of property initialization in classes. */
+    "noImplicitThis": true,                   /* Raise error on 'this' expressions with an implied 'any' type. */
+    "alwaysStrict": true,                     /* Parse in strict mode and emit "use strict" for each source file. */
```
Comment (liangyuanruo) [inline tsconfig.json:39]: `"strict": true` automatically marks all the other strict type checking options as true, so there's no need to specify them

File: test/server/api/cache/url.test.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
+import 'reflect-metadata'
+import { spy } from 'sinon'
```
Comment (liangyuanruo) [inline test/server/api/cache/url.test.ts:2]: Can I ask why not use Jest's [spyOn](https://jestjs.io/docs/en/jest-object.html#jestspyonobject-methodname)?

Comment (JasonChong96) [reply test/server/api/cache/url.test.ts:2]: Jest's spyOn doesn't seem to allow matching of only some arguments. In this case, we're only the function being spied on takes in a callback after the parameters and its only necessary to test that the correct parameters are being passed in.

Comment (liangyuanruo) [reply test/server/api/cache/url.test.ts:2]: just a query - no need to rework if it's too much work.

File: test/server/api/cache/url.test.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
+import 'reflect-metadata'
```
Comment (yong-jie) [inline test/server/api/cache/url.test.ts:1]: sorry to revive this PR, but how come every test file need to import this?

Comment (yong-jie) [reply test/server/api/cache/url.test.ts:1]: consider using jest's setupfile array that sets up everything nicely to reduce boilerplate https://jestjs.io/docs/en/configuration#setupfiles-array

Comment (liangyuanruo) [reply test/server/api/cache/url.test.ts:1]: good catch. can touch this up in a subsequent PR.

Comment (JasonChong96) [reply test/server/api/cache/url.test.ts:1]: Addressed in #84

**Pull request merged**